File: ios/Nishany/Core/Reader/AnnotationObject.swift
import Foundation

/// What a student writes on a document, and where it lives on the page.
///
/// A port of `src/lib/reader/annotations.ts`.
///
/// **Coordinates.** Everything is stored in *page space*: both axes divided by
/// the page's width in points. `x` runs 0..1; `y` runs 0..(height/width), about
/// 1.414 on A4. Dividing both by the same number is what keeps a circle round
/// and lets one divisor convert stroke widths too — so a mark survives zooming,
/// rotating the device, a different screen and a different pixel ratio, none of
/// which are recorded. Origin is the page's **top-left with y downward**, which
/// is the opposite of PDFKit; see `PageSpace` for the flip.
enum ObjectKind: String, Codable, Sendable {
    case ink, highlighter, tape, note, textbox, marker
}

/// The four pens.
///
/// Stored, but every one renders at flat width — the web's `widthProfile()`
/// exists and has no call site. Tapering here would make the same stroke look
/// different in a browser.
enum PenTool: String, Codable, Sendable {
    case ball, fountain, brush, pencil
}

/// Widget colours are theme tokens, not literal ink.
///
/// A sticky note is chrome — it has to stay readable on porcelain and on
/// charcoal — while a red pen has to stay red in both. The same eight tones as
/// the whiteboard, so a student meets one palette rather than two.
enum NoteTone: String, Codable, CaseIterable, Sendable {
    case paper, teal, amber, rose, sage, slate, sand, clay
}

/// One mark on a page.
///
/// A single struct rather than an enum with associated values, because the
/// stored form is one JSON object whose fields depend on `kind`, and a shard
/// containing a kind this build does not know must survive a round trip rather
/// than being dropped.
struct AnnotationObject: Codable, Identifiable, Equatable, Sendable {
    var id: String
    var kind: ObjectKind
    /// 1-based, matching everything else the reader says about pages.
    var page: Int
    /// Fractional, so inserting between two marks never renumbers a page.
    var z: Double
    /// `[x0, y0, x1, y1]` in page space, cached so culling and hit-testing can
    /// reject a mark without decoding its points.
    var bbox: [Double]
    /// Updated-at, ms since epoch. Must be strictly monotonic — the renderer
    /// keys off it to know a mark changed.
    var t: Double

    // Ink and highlighter
    /// `ball | fountain | brush | pencil | highlighter`.
    var tool: String?
    /// Literal sRGB: a red pen is red in every theme.
    var color: String?
    /// Base stroke width, in page-space units.
    var w: Double?
    /// 0..1. The highlighter is translucent, ink is not.
    var a: Double?
    /// Quantised, delta-encoded points.
    var p: [Int]?

    // Tape, note, textbox
    /// `[x0, y0, x1, y1]` in page space.
    var r: [Double]?
    var tone: NoteTone?
    var text: String?
    /// Page-space units, like a stroke width — so it scales with the page.
    var size: Double?

    // Marker
    var title: String?

    var isInk: Bool { kind == .ink || kind == .highlighter }

    /// The rectangle a boxed widget occupies, or the cached bounds otherwise.
    var rect: [Double] { r ?? bbox }
}

extension AnnotationObject {

    /// A fresh id.
    ///
    /// Ten lowercase alphanumerics, as the web generates: a UUID is 36 bytes
    /// and a book holds thousands of these.
    static func newID() -> String {
        let alphabet = Array("0123456789abcdefghijklmnopqrstuvwxyz")
        return String((0..<10).map { _ in alphabet[Int.random(in: 0..<alphabet.count)] })
    }

    /// A timestamp that is always greater than the last one used.
    ///
    /// The renderer treats `t` as "this changed", so two marks made inside the
    /// same millisecond must still be ordered.
    static func nextStamp(after previous: Double?) -> Double {
        let now = Date().timeIntervalSince1970 * 1000
        guard let previous else { return now.rounded() }
        return max(now.rounded(), previous + 1)
    }

    static func ink(
        kind: ObjectKind, tool: String, color: String, width: Double, alpha: Double?,
        points: [InkPoint], page: Int, z: Double, stamp: Double
    ) -> AnnotationObject {
        let simplified = StrokeCodec.simplify(points)
        return AnnotationObject(
            id: newID(), kind: kind, page: page, z: z,
            bbox: StrokeCodec.bounds(simplified, padding: width / 2),
            t: stamp,
            tool: tool, color: color, w: width, a: alpha,
            p: StrokeCodec.encode(simplified)
        )
    }

    static func widget(
        kind: ObjectKind, rect: [Double], tone: NoteTone?, text: String?, color: String?,
        size: Double?, page: Int, z: Double, stamp: Double
    ) -> AnnotationObject {
        AnnotationObject(
            id: newID(), kind: kind, page: page, z: z, bbox: rect, t: stamp,
            color: color, r: rect, tone: tone, text: text, size: size
        )
    }

    static func marker(title: String, page: Int, z: Double, stamp: Double) -> AnnotationObject {
        AnnotationObject(
            id: newID(), kind: .marker, page: page, z: z,
            bbox: [0, 0, 0, 0], t: stamp, title: title
        )
    }
}

/// What the manifest holds: the things that must be listable before the pages
/// they live on have loaded.
struct AnnotationManifest: Codable, Equatable, Sendable {
    var markers: [AnnotationObject] = []
    var notes: [NoteEntry] = []

    /// A searchable mirror of note and textbox text.
    ///
    /// Kept outside the shards so a search from page 1 can match a note on page
    /// 300 without loading twenty shards. Tape and ink are deliberately not
    /// indexed — neither carries text.
    struct NoteEntry: Codable, Equatable, Identifiable, Sendable {
        var id: String
        var page: Int
        var text: String
        /// `note` or `textbox`.
        var kind: String
    }

    /// Rebuild the index from a page's objects, keeping the web's ordering:
    /// notes by page then id, markers by page.
    static func entries(from objects: [AnnotationObject]) -> [NoteEntry] {
        objects
            .filter { ($0.kind == .note || $0.kind == .textbox) && !($0.text ?? "").isEmpty }
            .map { NoteEntry(id: $0.id, page: $0.page, text: $0.text ?? "", kind: $0.kind.rawValue) }
            .sorted { $0.page != $1.page ? $0.page < $1.page : $0.id < $1.id }
    }
}

File: android/app/src/main/java/com/nishany/android/core/reader/AnnotationObject.kt
package com.nishany.android.core.reader

import kotlinx.serialization.Serializable

/**
 * What a student writes on a resource, and where it lives on the page.
 *
 * A port of `src/lib/reader/annotations.ts` (mirrored by iOS
 * `Core/Reader/AnnotationObject.swift`).
 *
 * **Coordinates.** Everything is stored in *page space*: both axes divided by
 * the page's width in points. `x` runs 0..1; `y` runs 0..(height/width),
 * about 1.414 on A4. Dividing both by the same number is what keeps a circle
 * round and lets one divisor convert stroke widths too -- so a mark survives
 * zooming, resizing the window, a different screen, and a device with a
 * different pixel ratio, without any of them being recorded.
 *
 * **[kind] is a plain string, not a closed enum.** TypeScript's `ObjectKind`
 * union is a compile-time-only check -- nothing on the web validates it at
 * runtime -- so a kind this build has never heard of already passes straight
 * through JS untouched. A Kotlin enum would instead throw the moment it saw
 * one, corrupting the whole shard it lives in. This build only *renders*
 * `ink`/`highlighter`; everything else (`tape`, `note`, `textbox`, `marker`)
 * is decoded and re-encoded through the same flat set of optional fields
 * below, so a mark made on web or iOS is carried, unrendered, rather than
 * dropped -- see `AnnotationObjectTest` for the round-trip proof.
 *
 * A single flat class rather than a sealed hierarchy, for the same reason:
 * the stored JSON is one object whose fields depend on `kind`, and every
 * field any *known* kind uses is modelled here so a round trip through
 * [com.nishany.android.core.CortexJson] never has to drop one. The one edge
 * this does not cover -- a brand new kind introduced with a field name none
 * of the six existing kinds ever used -- would still be dropped by
 * [com.nishany.android.core.CortexJson]'s `ignoreUnknownKeys`, exactly as it
 * would for any other synced document in this app; that is an accepted,
 * documented ceiling (see `CortexJson`'s own doc), not one this model tries
 * to solve on its own.
 */
@Serializable
data class AnnotationObject(
    /** Short and random: a UUID is 36 bytes, and a book holds thousands. */
    val id: String,
    val kind: String,
    /** 1-based, matching everything else the reader says about pages. */
    val page: Int,
    /** Fractional, so inserting between two marks never renumbers a page. */
    val z: Double,
    /** `[x0, y0, x1, y1]` in page space. */
    val bbox: List<Double>,
    /** Updated-at, ms since epoch. Must be strictly monotonic. */
    val t: Double,

    // Ink and highlighter
    /** `ball | fountain | brush | pencil | highlighter`. */
    val tool: String? = null,
    /** Literal sRGB: a red pen is red in every theme. */
    val color: String? = null,
    /** Base stroke width, in page-space units. */
    val w: Double? = null,
    /** 0..1; the highlighter is translucent, ink is not. */
    val a: Double? = null,
    /** Quantised, delta-encoded points. See [StrokeCodec]. */
    val p: List<Int>? = null,

    // Tape, note, textbox (not authored or rendered by this build -- see class doc)
    /** `[x0, y0, x1, y1]` in page space. */
    val r: List<Double>? = null,
    val tone: String? = null,
    val text: String? = null,
    /** Page-space units, like a stroke width -- so it scales with the page. */
    val size: Double? = null,

    // Marker (not authored or rendered by this build -- see class doc)
    val title: String? = null,
) {
    val isInk: Boolean get() = kind == KIND_INK || kind == KIND_HIGHLIGHTER

    companion object {
        const val KIND_INK = "ink"
        const val KIND_HIGHLIGHTER = "highlighter"

        private val ID_ALPHABET = ('0'..'9') + ('a'..'z')

        /** Ten lowercase alphanumerics, matching the web's `newObjectId`. */
        fun newId(): String = (1..10).map { ID_ALPHABET.random() }.joinToString("")

        /** A timestamp always greater than [previous], so two marks in the same millisecond still order. */
        fun nextStamp(previous: Double?): Double {
            val now = System.currentTimeMillis().toDouble()
            return if (previous == null) now else maxOf(now, previous + 1)
        }

        /** A fresh ink or highlighter mark, points simplified and encoded exactly as [StrokeCodec] specifies. */
        fun ink(
            kind: String,
            tool: String,
            color: String,
            width: Double,
            alpha: Double?,
            points: List<InkPoint>,
            page: Int,
            z: Double,
            stamp: Double,
        ): AnnotationObject {
            val simplified = StrokeCodec.simplify(points)
            return AnnotationObject(
                id = newId(),
                kind = kind,
                page = page,
                z = z,
                bbox = StrokeCodec.bounds(simplified, padding = width / 2),
                t = stamp,
                tool = tool,
                color = color,
                w = width,
                a = alpha,
                p = StrokeCodec.encode(simplified),
            )
        }
    }
}
